Unresolved dependency warning includes source position. Fixes #528

eed3si9n opened the pull request:
Fixes #528.

Unresolved dependency warning is moved to `UnresolvedDependencyWarning` class including
the fail path that was added in #1467.
To display the source position, I need to access the State, so I had to move the
error processing out of `IvyActions` and add `UnresolvedDependencyWarning`, which is
aware of State.

Here's global plugin dependency sample:

```
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  ::          UNRESOLVED DEPENDENCIES         ::
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  :: com.typesafe.sbt#sbt-pgp;0.100.1: not found
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn] 
[warn]  Note: Some unresolved dependencies have extra attributes.  Check that these dependencies exist with the requested attributes.
[warn]      com.typesafe.sbt:sbt-pgp:0.100.1 (sbtVersion=0.13, scalaVersion=2.10)
[warn] 
[warn]  Note: Unresolved dependencies path:
[warn]      com.typesafe.sbt:sbt-pgp:0.100.1 (sbtVersion=0.13, scalaVersion=2.10) (/Users/foo/dotfiles/sbt/0.13/plugins/gpg.sbt#L3-4)
[warn]        +- org.scala-sbt:global-plugins:0.0 (sbtVersion=0.13, scalaVersion=2.10)
```

Normal library dependency sample:

```
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  ::          UNRESOLVED DEPENDENCIES         ::
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  :: foundrylogic.vpp#vpp;2.2.1: not found
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn] 
[warn]  Note: Unresolved dependencies path:
[warn]      foundrylogic.vpp:vpp:2.2.1
[warn]        +- org.apache.cayenne:cayenne-tools:3.0.2
[warn]        +- org.apache.cayenne.plugins:maven-cayenne-plugin:3.0.2 (/Users/foo/hello/build.sbt#L28)
[warn]        +- d:d_2.10:0.1-SNAPSHOT
```

/review @jsuereth /cc @gkossakowski

eed3si9n committed:
Unresolved dependency warning includes source position. Fixes #528

Unresolved dependency warning is moved to UnresolvedDependencyWarning class including
the fail path that was added in #1467.
To display the source position, I need to access the State, so I had to move the
error processing out of IvyActions and add UnresolvedDependencyWarning, which is
aware of State.

jsuereth commented on main/src/main/scala/sbt/UnresolvedDependencyWarning.scala:
```diff
@@ -0,0 +1,81 @@
+          val empty = extracted.structure.data set (sk.scope, sk.key, Nil)
+          val settings = extracted.structure.settings filter { s: Setting[_] =>
+            (s.key.key == Keys.libraryDependencies.key) &&
+              (s.key.scope.project == Select(extracted.currentRef))
```
Current ref may not be the reference to the project in which update is running.  Have you tried this on a multi-project build?

jsuereth commented on main/src/main/scala/sbt/UnresolvedDependencyWarning.scala:
```diff
@@ -0,0 +1,81 @@
+    // This is a mapping between modules and original position, in which the module was introduced.
+    lazy val modulePositions: Seq[(ModuleID, SourcePosition)] =
+      try {
+        stateOpt map { state =>
```
Why don't you pass all this in via some "config" option rather than via state?

jsuereth commented:
So, this looks clever and nice for the user.  Major concern is your use of projectRef.   I believe there's a way, inside a setting, to grab the scope of the setting itself (see how the sourceDirectory settings are generated).   The way you're currently using state is a bit suspect though, but generally I think it's the only way to grab the information you want.

What may be nicer, architecturally, is some kind of state post-process step which auto-generates a "LibraryDependencySourceIndex" setting you can just depend on.   This setting could hide the complexities of looking up source positions.

Ideally, we don't want the cost of scanning through settings to be hit EVERY TIME you run resolution, so we may even want to try to cache such an index after load time and avoid altering it until the libraryDependencies settings change.

eed3si9n committed:
Adds UnresolvedWarningConfiguration that caches ModuleID -> SourcePosition mapping.

UnresolvedWarning is moved back to IvyActions.scala where it belongs.
The mapping between ModuleID and SourcePosition is passed in as UnresolvedWarningConfiguration.
This is calculated once in Defaults using State and is cached to filesystem.

eed3si9n commented:
> Major concern is your use of projectRef.

You are right about not trusting `extracted.currentRef`. I'm now using `thisProjectRef` instead, and that one seems to work.

> What may be nicer, architecturally, is some kind of state post-process step which auto-generates a "LibraryDependencySourceIndex" setting you can just depend on.

Agreed. With the new design, that `Module -> SourcePosition` index is called `UnresolvedWarningConfiguration`, which is calculated using a new task `unresolvedWarningConfiguration`. 

> Ideally, we don't want the cost of scanning through settings to be hit EVERY TIME you run resolution, so we may even want to try to cache such an index after load time and avoid altering it until the libraryDependencies settings change.

`unresolvedWarningConfigurationTask` does just that:

``` scala
    val f = Tracked.inputChanged(cacheFile / "input_uwc") { (inChanged: Boolean, in: Seq[ModuleID]) =>
      val outCache = Tracked.lastOutput[Seq[ModuleID], UnresolvedWarningConfiguration](outCacheFile) {
        case (_, Some(out)) if !inChanged => out
        case _                            => UnresolvedWarningConfiguration(modulePositions)
      }
      outCache(in)
    }
    f(libraryDependencies.value)
```

jsuereth commented:
So now may only complaint is that `UnresolvedWarningConfiguration` is a name about what you're trying to do but not what the configuration actually IS.  It's not actually the configuration for unresolved warnings, but a "LibraryDependencyDefinitionIndex" or some such.  Other than petty naming discussions, LGTM.

eed3si9n committed:
Adds `dependencyPositions` task to explicitly track dependency positions.

eed3si9n commented:
Added `dependencyPositions` task to cache dependency positions explicitly.

jsuereth commented:
Cool.  LGTM
